pkg/kernel/factory.go

```diff
@@ -0,0 +1,173 @@
+package kernel
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/justtrackio/gosoline/pkg/cfg"
+	"github.com/justtrackio/gosoline/pkg/coffin"
+	"github.com/justtrackio/gosoline/pkg/log"
+)
+
+type factory struct {
+	ctx       context.Context
+	config    cfg.Config
+	logger    log.Logger
+	blueprint *blueprint
+
+	middlewares []Middleware
+	stages      stages
+}
+
+func NewFactory(ctx context.Context, config cfg.Config, logger log.Logger, bp *blueprint) (*factory, error) {
+	factory := &factory{
+		ctx:       ctx,
+		config:    config,
+		logger:    logger.WithChannel("kernel"),
+		blueprint: bp,
+
+		middlewares: make([]Middleware, 0),
+		stages:      make(stages),
+	}
+
+	if err := factory.build(); err != nil {
+		return nil, fmt.Errorf("can not build kernel factory: %w", err)
+	}
+
+	return factory, nil
+}
+
+func (f *factory) GetKernel() (Kernel, error) {
```

Comment:
This function never returns an error, why does the caller have to check for one?

Reply:
changed